download-data.py

```diff
@@ -4,20 +4,150 @@
+    ]
+
+
+def get_results_modules(bucket: str, release: str, profile: str) -> List[str]:
+    """
+    Get the list of available results modules from the OpenScPCA results bucket.
+    """
+    ls_cmd = ["aws", "s3", "ls", f"s3://{bucket}/{release}/"]
+    if profile:
+        ls_cmd += ["--profile", profile]
+    if bucket == TEST_BUCKET:
+        ls_cmd += ["--no-sign-request"]
+    ls_result = subprocess.run(ls_cmd, capture_output=True, text=True)
+    if ls_result.returncode:
+        print(
+            "Error listing results modules from the OpenScPCA results bucket.",
+            " Ensure you have the correct AWS permissions to access OpenScPCA data.",
+            file=sys.stderr,
+        )
+    ls_result.check_returncode()
+
+    # get only prefixes and remove the trailing slash
+    module_re = re.compile(r"PRE\s+([\S]+)/")
+    return [
+        m.group(1)
+        for m in (module_re.search(line) for line in ls_result.stdout.splitlines())
+        if m
+    ]
+
+
+def build_sync_cmd(
+    bucket: str,
+    release: str,
+    download_dir: pathlib.Path,
+    include_patterns: List[str] = [],
+    dryrun: bool = False,
+    profile: str = "",
+) -> List[str]:
+    sync_cmd = [
+        "aws",
+        "s3",
+        "sync",
+        f"s3://{bucket}/{release}/",
+        download_dir,
+        "--exact-timestamps",  # replace if a file has changed at all
+        "--no-progress",  # don't show progress animations
+    ]
+
+    if include_patterns:
+        sync_cmd += ["--exclude", "*"]
+
+    for pattern in include_patterns:
+        sync_cmd += ["--include", pattern]
+
+    if dryrun:
+        sync_cmd += ["--dryrun"]
+
+    if profile:
+        sync_cmd += ["--profile", profile]
+
+    if bucket == TEST_BUCKET:
+        sync_cmd += ["--no-sign-request"]
+
+    return sync_cmd
+
+
+def get_download_size(
+    bucket: str,
+    release: str,
+    include_patterns: List[str] = ["*"],
+    profile: str = "",
+) -> int:
+    """
+    Get the total size of files that will be downloaded from AWS S3 that match the include patterns.
+    """
+    ls_cmd = ["aws", "s3", "ls", f"s3://{bucket}/{release}/", "--recursive"]
+    if profile:
+        ls_cmd += ["--profile", profile]
+    if bucket == TEST_BUCKET:
+        ls_cmd += ["--no-sign-request"]
+    file_list = subprocess.run(ls_cmd, capture_output=True, text=True)
+    file_list.check_returncode()
+
+    total_size = 0
+    for line in file_list.stdout.splitlines():
+        size, file = line.split()[-2:]
+        if any(fnmatch.fnmatch(file, pattern) for pattern in include_patterns):
+            total_size += int(size)
+    return total_size
+
+
+def make_size_human(size: int) -> str:
+    """
+    Convert a size in bytes to something human readable.
+    """
+    for unit in ["B", "KiB", "MiB", "GiB", "TiB"]:
+        if size < 1024.0 or unit == "TiB":
+            break
+        size /= 1024.0
+    if unit == "B":
+        return f"{size:.0f} B"
+    return f"{size:.2f} {unit}"
```

Comment:
I don't think this is working as you expect? When I try to download the `merge-sce` results module with dry run I see a total size of 0 B, but the files are being listed.

Reply:
Update, if I include a `--projects` argument then the size is correct. So this doesn't work if you want to download results for an entire module.

Reply:
Hmm...I am presumably doing something wrong with the globs, since I am trying to match the behavior of `aws`. Will look at this case!

Reply:
I believe I fixed this in f1442f3. If you want to test again, that would be great! Thank you for the detailed report, it made it easy to find my error!

Reply:
Yes that fix works! I can now run a dry run for an entire module or a specific project within a module and the appropriate files are listed and the size is listed and appears to be correct. 🎉